src/script/lua_api/l_ipc.cpp

```diff
@@ -0,0 +1,138 @@
+// Minetest
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#include "lua_api/l_ipc.h"
+#include "lua_api/l_internal.h"
+#include "common/c_packer.h"
+#include "server.h"
+#include "debug.h"
+#include <chrono>
+
+typedef std::shared_lock<std::shared_mutex> SharedReadLock;
+typedef std::unique_lock<std::shared_mutex> SharedWriteLock;
+
+static inline auto read_pv(lua_State *L, int idx)
+{
+	std::unique_ptr<PackedValue> ret;
+	if (!lua_isnil(L, idx)) {
+		ret.reset(script_pack(L, idx));
+		if (ret->contains_userdata)
+			throw LuaError("Userdata not allowed");
+	}
+	return ret;
+}
+
+int ModApiIPC::l_ipc_get(lua_State *L)
+{
+	auto *store = getGameDef(L)->getModIPCStore();
+
+	auto key = readParam<std::string>(L, 1);
+
+	{
+		SharedReadLock autolock(store->mutex);
+		auto it = store->map.find(key);
+		if (it == store->map.end())
+			lua_pushnil(L);
+		else
+			script_unpack(L, it->second.get());
+	}
+	return 1;
+}
+
+int ModApiIPC::l_ipc_set(lua_State *L)
+{
+	auto *store = getGameDef(L)->getModIPCStore();
+
+	auto key = readParam<std::string>(L, 1);
+
+	luaL_checkany(L, 2);
+	auto pv = read_pv(L, 2);
+
+	{
+		SharedWriteLock autolock(store->mutex);
+		if (pv)
+			store->map[key] = std::move(pv);
+		else
+			store->map.erase(key); // delete the map value for nil
+		store->signal();
```

Comment:
the signal does not need to be done under lock, and (probably) shouldn't be:

https://en.cppreference.com/w/cpp/thread/condition_variable_any/notify_all

>The notifying thread does not need to hold the lock on the same mutex as the one held by the waiting thread(s). Doing so may be a pessimization, since the notified thread would immediately block again, waiting for the notifying thread to release the lock, though some implementations recognize the pattern and do not attempt to wake up the thread that is notified under lock.

Reply:
huh, this is different semantics from `pthread_cond_signal`

Reply:
Performance-wise maybe (?) it seems, but not when it comes to correctness? https://linux.die.net/man/3/pthread_cond_signal

> The pthread_cond_broadcast() or pthread_cond_signal() functions may be called by a thread whether or not it currently owns the mutex that threads calling pthread_cond_wait() or pthread_cond_timedwait() have associated with the condition variable during their waits; however, if predictable scheduling behavior is required, then that mutex shall be locked by the thread calling pthread_cond_broadcast() or pthread_cond_signal(). 

Not sure if predictable scheduling here just means predictable, or if it means faster. Or how it would work exactly if the broadcasting thread still holds a lock.